### DISCUSSION TITLE
Optimise infinite scroller

### DISCUSSION THREAD

**Discussion by danielniccoli** [alpinejs/alpine#3151] [1. Help] [answered]
I wrote an infinite scroll component that fetches more entries from a REST endpoint when the user almost reaches the end of the page. I am wondering if the `x-for` element re-renders the entire `entries` array after it gets updated, or if there is some intelligence behind it that only renders new array items. But I suspect the entire array gets rerendered. If so, do you have any idea how to optimise this?

```html
<main x-data="{ endpoint: '/api/v1/entries', entries: [] }">
    <ul>
        <template x-for="entry in entries">
            <li :id="entry.id">
                <article x-text="entry.title"></article>
            </li>
        </template>
    </ul>
    <div id="endofpage"
         x-intersect.margin.200px="
             result = await (await fetch(endpoint)).json();
             endpoint = result.next;
             entries = entries.concat(result.results)"/>
</main>
```

**Comment by ekwoka** [comment, marked as answer]
It generally shouldn't.

However, it's recommended to add a `:key` property to the template so that Alpine can correctly identify which element belongs to which item of the loop.

**Reply by danielniccoli** [reply]
When you say "It shouldn't" do you mean it should not re-render all templates for the entire array, but only the templates for the new array items?

I added the key, thanks for the hint!

**Reply by ekwoka** [reply]
Yes, it should simply swap any data that changes, and if nothing changes it won't do anything.

However, on updates it will do some processing to determine those differences.

Adding the key can help short circuit that check so that Alpine knows if the object changed and won't need to check the Dom as closely.

This does mean that as the number of items grows to very large amounts, there can be some overhead during array updates, but most of the cost will still be in the creation of the new elements.
